### PR TITLE
`allow_reentrancy` attribute

### PR DESCRIPTION
**Pull request by Artemka374**
Continuation of work on https://github.com/paritytech/ink/pull/1507
which is a part of https://github.com/paritytech/ink/issues/1285

### CODE REVIEW

File: .idea/.gitignore
Comment (ascjones) [inline .idea/.gitignore:1]: Please remove the `.idea` directory from the PR. You can ignore it locally with `echo .idea > .git/info/exclude`

Comment (Artemka374) [reply .idea/.gitignore:1]: done :)

**Comment by ascjones** [comment]
Merge `master` into this PR to fix the CI

File: crates/env/src/engine/off_chain/impls.rs [hunk 1]
```diff
@@ -550,4 +550,11 @@ impl TypedEnvBackend for EnvInstance {
     {
         unimplemented!("off-chain environment does not support `call_runtime`")
     }
+
+    fn reentrance_count<E>(&mut self) -> u32
+    where
+        E: Environment,
+    {
+        unimplemented!("off-chain environment does not support `reentrance_count`")
```
[the remark this answers is not in the document]
Comment (Artemka374) [reply crates/env/src/engine/off_chain/impls.rs:553]: I am not sure, that we should implement it in this PR, along with the fact, that we need to expand ExecContext in off-chain engine, so it looks for me as a point for another PR, maybe something like this: https://github.com/paritytech/ink/pull/1589

**Comment by deuszx** [comment]
What needs to be done here to get it merged? I can help with the coding work.

File: crates/env/src/backend.rs [hunk 2]
```diff
@@ -162,8 +173,8 @@ impl CallFlags {
     /// # Note
     ///
     /// See [`Self::set_allow_reentry`] for more information.
-    pub const fn allow_reentry(&self) -> bool {
-        self.allow_reentry
+    pub const fn deny_reentry(&self) -> bool {
```
[the remark this answers is not in the document]
Comment (SkymanOne) [reply crates/env/src/backend.rs:176]: Agreed. I don't see much point here too.

File: crates/ink/codegen/src/generator/dispatch.rs [hunk 3]
```diff
@@ -913,6 +959,36 @@ impl Dispatch<'_> {
         )
     }
 
+    fn any_message_accepts_reentrancy(
```
Comment (SkymanOne) [inline crates/ink/codegen/src/generator/dispatch.rs:962]: Please add docs here

File: crates/ink/codegen/src/generator/dispatch.rs [hunk 4]
```diff
@@ -948,4 +1024,31 @@ impl Dispatch<'_> {
             }
         )
     }
+
+    fn any_constructor_accepts_reentrancy(
```
Comment (SkymanOne) [inline crates/ink/codegen/src/generator/dispatch.rs:1028]: Add docs here too

File: crates/ink/ir/src/ir/item_impl/message.rs [hunk 5]
```diff
@@ -100,6 +100,8 @@ pub struct Message {
     pub(super) item: syn::ImplItemFn,
     /// If the ink! message can receive funds.
     is_payable: bool,
+    /// If the ink! message is allowed to re-enter the contract.
```
Comment (SkymanOne) [inline crates/ink/ir/src/ir/item_impl/message.rs:103]: Need some elaboration

File: crates/metadata/src/specs.rs [hunk 6]
```diff
@@ -621,6 +658,8 @@ pub struct MessageSpec<F: Form = MetaForm> {
     mutates: bool,
     /// If the message accepts any `value` from the caller.
     payable: bool,
+    /// If the message is allowed to re-enter the contract.
+    allow_reentrancy: bool,
```
Comment (SkymanOne) [inline crates/metadata/src/specs.rs:662]:
```suggestion
   reentrancy_allowed: bool,
```

**Review by SkymanOne — COMMENTED** [review]
Looks quite good for the first iteration. However, I am not fun of renaming `set_allow_reentry()`

**Comment by SkymanOne** [comment]
Closing due to inactivity

**Pull request closed without merging**